### ISSUE TITLE
Add optional string argument to line2byte(), and byte2line()

### ISSUE THREAD

**Issue by poetaman** [vim/vim#8206]
[ripgrep --json](https://docs.rs/grep-printer/0.1.5/grep_printer/struct.JSON.html) (apparently fastest regex engine out there) provides enough information to decipher ``lnum``, ``col``, ``end_lnum``, and ``end_col`` for search matches. That can be used with vim's text properties feature to highlight all the regex matches.

Though the nature of information provided requires one to implement equivalents of ``line2byte()`` and ``byte2line()`` that operate on text of interest provided by ripgrep's output instead of on the entire buffer . Currently ``line2byte()`` and ``byte2line()`` accept 1 argument, and operate on the entire text in a vim buffer. The feature request is to add an optional second argument ``{expr}`` (of type string). When ``{expr}`` is provided, the functions ``line2byte()``/``byte2line()`` will operate on text in ``{expr}`` instead of text of buffer.

Advantage/alternative: Ripgrep's json output provides (sub)match's start/end byte offset relative to the first line of match, but not the actual line numbers or byte offset (from start of respective line). It also provides the line number & absolute byte offset of first line of match. Note: this "first line of match" is just a reference line in a multi-line region where actual matches (called "submatches") can be anywhere. One can use existing functions ``line2byte()``/``byte2line()`` (without any modification), and absolute byte offsets to determine the line number and column number where a submatch starts/ends to apply highlighting. But that means always calculating from line 1, even though we just need to operate on few lines worth of "match text" that ripgrep provides. For long files this can be significantly slow.

> As a sidenote, ``ripgrep --json`` is the default search engine of [VS Code](https://code.visualstudio.com) according to the author [here](https://github.com/BurntSushi/ripgrep/issues/1866).

The following ascii art shows the nature of information provided by ``ripgrep --json``:

```
.
.
1334 -  -  -  -  -  -  -  -  x  x  x  x  x  x  
1335 x  x  x  x  -  -  -  -  -  -  -  -  -  -  -  x  x  x  x  x  
1336 x  x  x  x  x  x  x  x  x  x  x    
1337 x  x  x  x  x  x  x  x  x  x  x  x  x  x  x  x  x  
1338 x  x  x  x  x  x  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  -  
.
.
```

``ripgrep --json`` provides:
1. entire lines of a match region (from line 1334 to 1338) provided as "text" field;
2. line number, and absolute byte offset for first line of match region (line 1334);
3. an array with multiple "submatches", these subregions are marked with x's in the ascii above;
4. start/end byte offsets for all submatches relative to start of first line of match region (i.e beginning of 1334). We can use this to get absolute byte offsets to use with existing functions ``line2byte()``/``byte2line()`` and calculate start/end line numbers of subregions in a slow fashion. Or we can have extra argument to ``line2byte()``/``byte2line()`` that takes the entire match region (string that has lines 1334-1338) as a second argument, and only calculate line number offsets relative to 1334 (a line number we already know), etc.

**Comment by brammool** [comment]
Can you give the literal output that you need to work with?
I'm not sure whether using a string argument on byte2line() would be faster than using the buffer from the start.
For a buffer the size is "cached".  For a text string it would need to find line breaks and strlen(), which probably isn't faster.
We probably should use different functions anyway, since line2byte() and byte2line() are always working on buffers.

**Comment by poetaman** [comment]
@brammool Yes, I have it below. Forgot to mention, another more important reason (other than suspected perf disadvantage) is the timing of calculation of lnum, and col. I intend to feed the information from json into quickfix list. This ideally means that I calculate lnum and col before loading the files that had ``ripgrep`` matches into vim buffers. So thats another reason I cannot use the current interface of ``line2byte()`` and ``byte2line()``, as it only works on loaded vim buffers. But given ``ripgrep`` can match hundreds, or even thousands of files, ``For a buffer the size is "cached"`` doesn't help as user might never intend to open all of the hundreds/thousands of files. 

> As a sidenote, ripgrep provides a --vimgrep printer option that can be directly used to feed quickfix list. But the limitation of that format is that it doesn't print the end_lnum, end_col (something that can only be calculated from its --json printer output)

Now the literal output, input and command to generate the information:

1) Save file ``sample.tex``:

```tex
This is some
random text
before the match

\setmainfont{LucidaBrightOT}[
    Extension = .otf,
    Path = fonts/LucidaSans/,
    UprightFont = *,
    BoldFont = *-Demi,
    ItalicFont = *-Italic,
    BoldItalicFont = *-DemiItalic
] And the parantheses (have some
information that gets

captured ) but nothing here

And this is some more 
random text 
after the match
```


2) Run ripgrep command: ``>>rg --json -U '(?s)\[.*\]|\(.*\)|\{.*\}' sample.tex``

3) Pretty print one of the json's generated with ``"type": "match"``. I have some comments in json below that are marked with ``<--``

```json
{
  "type": "match",
  "data": {
    "path": {
      "text": "sample.tex"
    },
    "lines": {
      "text": "\\setmainfont{LucidaBrightOT}[\n    Extension = .otf,\n    Path = fonts/LucidaSans/,\n    UprightFont = *,\n    BoldFont = *-Demi,\n    ItalicFont = *-Italic,\n    BoldItalicFont = *-DemiItalic\n] And the parantheses (have some\ninformation that gets\n\ncaptured ) but nothing here\n"
    }, <-- lines.text has contents of all the lines in match region (lines 1334-1338 from ticket's ascii art)
    "line_number": 5, <-- line number of beginning of match region (line number 1334 from ticket's ascii art)
    "absolute_offset": 43, <-- absolute byte offset of match region from beginning of file
    "submatches": [
      {
        "match": {
          "text": "{LucidaBrightOT}"
        },
        "start": 12, <-- start byte offset of a submatch relative to beginning of match region
        "end": 28 <--  end byte offset of a submatch relative to beginning of match region
      },
      {
        "match": {
          "text": "[\n    Extension = .otf,\n    Path = fonts/LucidaSans/,\n    UprightFont = *,\n    BoldFont = *-Demi,\n    ItalicFont = *-Italic,\n    BoldItalicFont = *-DemiItalic\n]"
        },
        "start": 28, <-- start byte offset of a submatch relative to beginning of match region
        "end": 188  <--  end byte offset of a submatch relative to beginning of match region
      },
      {
        "match": {
          "text": "(have some\ninformation that gets\n\ncaptured )"
        },
        "start": 209, <--  start byte offset of a submatch relative to beginning of match region
        "end": 253  <--  end byte offset of a submatch relative to beginning of match region
      }
    ]
  }
}
```

I intend to feed ``data.lines.text`` as a second argument to ``line2byte()`` and ``byte2line()`` and use it along side the ``start``/``end`` byte offsets of submatches. The only other option seems to use ``strpart()``, and calculate line breaks, and what not... It essentially will replicate the functionality of ``line2byte()`` and ``byte2line()`` to operate on string ``data.lines.text`` from json. So I thought if the functionality is already implemented in those functions, adding an optional argument can expose it to users.

**Comment by bfrg** [comment]
@reportaman Just FYI, if you use `javascript` code blocks, you can use C-style comments without getting those red error lines. For example:

```javascript
{
  "type": "match",
  "data": {
    "path": {
      "text": "sample.tex"
    },
    "lines": {
      // lines.text has contents of all the lines in match region (lines 1334-1338 from ticket's ascii art)
      "text": "\\setmainfont{LucidaBrightOT}[\n    Extension = .otf,\n    Path = fonts/LucidaSans/,\n    UprightFont = *,\n    BoldFont = *-Demi,\n    ItalicFont = *-Italic,\n    BoldItalicFont = *-DemiItalic\n] And the parantheses (have some\ninformation that gets\n\ncaptured ) but nothing here\n"
    },
    "line_number": 5,      // line number of beginning of match region (line number 1334 from ticket's ascii art)
    "absolute_offset": 43, // absolute byte offset of match region from beginning of file
    "submatches": [
      {
        "match": {
          "text": "{LucidaBrightOT}"
        },
        "start": 12, // start byte offset of a submatch relative to beginning of match region
        "end": 28    // end byte offset of a submatch relative to beginning of match region
      },
      {
        "match": {
          "text": "[\n    Extension = .otf,\n    Path = fonts/LucidaSans/,\n    UprightFont = *,\n    BoldFont = *-Demi,\n    ItalicFont = *-Italic,\n    BoldItalicFont = *-DemiItalic\n]"
        },
        "start": 28, // start byte offset of a submatch relative to beginning of match region
        "end": 188   // end byte offset of a submatch relative to beginning of match region
      },
      {
        "match": {
          "text": "(have some\ninformation that gets\n\ncaptured )"
        },
        "start": 209, // start byte offset of a submatch relative to beginning of match region
        "end": 253    // end byte offset of a submatch relative to beginning of match region
      }
    ]
  }
}
```

**Comment by poetaman** [comment]
@bfrg Ok thanks! Will keep in mind for future.

**Comment by brammool** [comment]
> @brammool Yes, I have it below. Forgot to mention, another more important reason (other than suspected perf disadvantage) is the timing of calculation of lnum, and col. I intend to feed the information from json into quickfix list. This ideally means that I calculate lnum and col before loading the files that had ``ripgrep`` matches into vim buffers. So thats another reason I cannot use the current interface of ``line2byte()`` and ``byte2line()``, as it only works on loaded vim buffers. But given ``ripgrep`` can match hundreds, or even thousands of files, ``For a buffer the size is "cached"`` doesn't help as user might never open some files.
> 
> Now the literal output, input and command to generate the information:
> 
> 1) Save file ``sample.tex``:
> 
> ```tex
> This is some
> random text
> before the match
> 
> \setmainfont{LucidaBrightOT}[
>     Extension = .otf,
>     Path = fonts/LucidaSans/,
>     UprightFont = *,
>     BoldFont = *-Demi,
>     ItalicFont = *-Italic,
>     BoldItalicFont = *-DemiItalic
> ] And the parantheses (have some
> information that gets
> 
> captured ) but nothing here
> 
> And this is some more 
> random text 
> after the match
> ```
> 
> 
> 2) Run ripgrep command: ``>>rg --json -U '(?s)\[.*\]|\(.*\)|\{.*\}' sample.tex``
> 
> 3) Pretty print one of the json's generated with ``"type": "match"``. I have some comments in json below that are marked with ``<--``
> 
> ```json
> {
>   "type": "match",
>   "data": {
>     "path": {
>       "text": "sample.tex"
>     },
>     "lines": {
>       "text": "\\setmainfont{LucidaBrightOT}[\n    Extension = .otf,\n    Path = fonts/LucidaSans/,\n    UprightFont = *,\n    BoldFont = *-Demi,\n    ItalicFont = *-Italic,\n    BoldItalicFont = *-DemiItalic\n] And the parantheses (have some\ninformation that gets\n\ncaptured ) but nothing here\n"
>     }, <-- lines.text has contents of all the lines in match region (lines 1334-1338 from ticket's ascii art)
>     "line_number": 5, <-- line number of beginning of match region (line number 1334 from ticket's ascii art)
>     "absolute_offset": 43, <-- absolute byte offset of match region from beginning of file
>     "submatches": [
>       {
>         "match": {
>           "text": "{LucidaBrightOT}"
>         },
>         "start": 12, <-- start byte offset of a submatch relative to beginning of match region
>         "end": 28 <--  end byte offset of a submatch relative to beginning of match region
>       },
>       {
>         "match": {
>           "text": "[\n    Extension = .otf,\n    Path = fonts/LucidaSans/,\n    UprightFont = *,\n    BoldFont = *-Demi,\n    ItalicFont = *-Italic,\n    BoldItalicFont = *-DemiItalic\n]"
>         },
>         "start": 28, <-- start byte offset of a submatch relative to beginning of match region
>         "end": 188  <--  end byte offset of a submatch relative to beginning of match region
>       },
>       {
>         "match": {
>           "text": "(have some\ninformation that gets\n\ncaptured )"
>         },
>         "start": 209, <--  start byte offset of a submatch relative to beginning of match region
>         "end": 253  <--  end byte offset of a submatch relative to beginning of match region
>       }
>     ]
>   }
> }
> ```

It appears you need to take the text, split it into lines with split()
and then use the length of each line.  You can do that with Vim script.
This might be a bit slow if you have many matches, thus you would need
to try to postpone it.  You could add the matches at first with the
starting line number only.  Then when a buffer is loaded, look into the
matches for that buffer and fill in the exact positions.  You can use
getqflist() and setqflist().

-- 
Ten bugs in the hand is better than one as yet undetected.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by poetaman** [comment]
@brammool Yep, but the "lnum" & "col" is going to be incorrect for submatches down in the match region, and quickfix list will hold wrong information. Currently I am using ``strpart()`` and ``split()`` to find number of ``\n``'s, etc. What are your thoughts on my proposal? Because I can avoid doing the slow calculation you mention if ``line2byte()`` and ``byte2line()`` could accept a string (in my case of the match region).
